message.go

```diff
@@ -4,54 +4,246 @@ import (
 	"errors"
 )
 
-type message struct {
-	method          string
-	seqno           seqNumber
-	res             interface{}
-	err             interface{}
-	remainingFields int
-	decodeSlots     []interface{}
-}
-
-func decodeIntoMessage(dec decoder, m *message) error {
-	for _, s := range m.decodeSlots {
-		if err := decodeMessage(dec, m, s); err != nil {
-			return err
+type RPCMessage interface {
+	Type() MethodType
```

Comment:
I think you can get away with not having this function -- the only non-test implementation is RPCCall, and you just do

```
+	rpc := &RPCCall{}
 +	err = rpc.Decode(nb-0x90, p.frameDecoder, p.protocols, p.calls)
```

so you can turn rpc.Decode back into a free function and move Type() into RPCData. This also gets rid of the weird circular dependency with ResponseCh returning an RPCMessage instead of RPCData.